### DISCUSSION TITLE
Best approach to computed fields in a entity hierarchy

### DISCUSSION THREAD

**Discussion by jsboige** [2sic/2sxc#2350] [Q&A]
Given a hierarchy of entities with a child->parent.Guid field to  implement a m-1 relationship, what is the best approach to retrieve computed fields, namely ancestor parents of level 1-->3, and have them readily available together with other entity fields?
Are DataType: "Entity", and Data-input "Query-based entity" a good pick for those fields?

**Comment by iJungleboy** [comment]
Hi @jsboige - your question is described so generically that I have no idea what you mean 😂

Can you give an example?

**Reply by jsboige** [reply]
I'm working with a taxonomy of "fallacy" entities. I believe we discussed briefly that project. Each fallacy can have a parent, starting from a root node and then 7 famillies and subsequent sub-trees.
The Parent field defined as a related entity works fine, and I was able to import the existing csv data through Xml exports using that model. Now I'd like to have available Family/ Subfamily/SubsubFamily as computed fields readily available for each Fallacy entity. The fields were defined within the Fallacy entity type as Query-based, then I'm not sure how to proceed from there.

**Comment by iJungleboy** [comment]
If I understand you correctly, you already have the data in a multi-level tree in the entities, and you are asking how to access these children and grandchildren in code for display or use in LINQ. Is this correct? So something like

1. Grand-Parent Item (grand-parent)
    1. Parent Item (parent)
         1. Current Item
             1. Child Item 1
             2. Child Item 2
             3. Child Item 3
                  1. Grandchild 1
                  1. Grandchild 2

Is this correctly describing your need?

**Reply by jsboige** [reply]
Your picture is correct. 
For each node, we would need parent, children, and the top 3 ancestors (level 1,2,3 of the tree, discarding the root).
I suppose parent and children access are standard needs, and I'll try to find examples, in the "Events" app maybe?
"Top 3 ancestors" is more of a specific need, and the one I was inquiring about.

**Reply by iJungleboy** [reply]
Navigating relationships works as follows:

`thing.FieldWithEntities`  gives you an IEnumerable<dynamic> of the child-items in that field. Razor will have some difficulty with LINQ when you want to do things like `thing.FieldWithEntities.Any()` because the compiler can't detect that `FieldWithEntities` is an IEnumerable, so for doing a foreach or similar it's usually best to do `AsList(thing.FieldWithEntities)` or in some cases you need to also use `as object` meaning you'll have `AsList(thing.FieldWithEntities as object)` and then you can enumerate it. 

Parents are of course not a field on this object, so you'll use the Parents accessor `thing.Parents("typeNameOfParent")`. This returns the same signature, but Razor may also need the `AsList(... as object)` to help the compiler figure things out. 

There are a few tricks more to `.Parents`, best check out the docs https://docs.2sxc.org/api/dot-net/ToSic.Sxc.Data.DynamicEntity.html#ToSic_Sxc_Data_DynamicEntity_Parents_System_String_System_String_

There is also a corresponding `.Children` https://docs.2sxc.org/api/dot-net/ToSic.Sxc.Data.DynamicEntity.html#ToSic_Sxc_Data_DynamicEntity_Children_System_String_System_String_

With this, you can do a lot of stuff, for example:

```
// gets you an IEnumerable<dynamic> of all grand-parents
var allGrandparents = AsList(thing.Parents("ParentType") as object)
                                .SelectMany(p => AsList(p.Parents("GrantparentType") as object));
```

You'll find more examples here https://2sxc.org/dnn-tutorials/en/razor/linq/home

**Reply by jsboige** [reply]
"Parent" is the entity field that currently allows defining the hierarchy (as with DNN pages for instance), so I have the impression that you got the model upside down.
As I understand it, I can't really define the Family/Subfamily fields to be computed at runtime, and should instead define Razor helpers to make that data available from the UI without the content type definition acknowledging their existence. About those helpers, would it be possible to compile a project referencing 2sxc to manipulate the dynamic entities directly or does it have to be within interpreted razor views?

**Reply by iJungleboy** [reply]
I don't really understand the questions :), but I'll try to answer anyhow. 

An entity (Book) with a field Author sees Author as a child. The author-entity can find the book with Parent. You can use either relationship navigation that you need. 

Assuming you want to have a DLL which has some business logic working with the data, then you will usually not work with `dynamic` data types, but with typed data. You could work with `dynamic` but it just increases the risk of typos, so I don't recommend it. If you will work with typed data, then you have two options:

1. Using the normal Entity objects (so no `AsyDynamic(...)`) https://docs.2sxc.org/api/dot-net/ToSic.Eav.Data.IEntity.html - this also has Parents / Children accessors, and to get a field value you would usually use `.Value<T>("name")` or `.GetBestValue<T>("name", optional-list-of-languages)`
2. Create strongly typed wrapper objects for your data, using EntityBasedTypes https://docs.2sxc.org/api/dot-net/ToSic.Eav.Data.EntityBasedType.html - in this case you would create an class called `Book` with the properties you want, inheriting from `EntityBasedType` or the logging variant `EntityBasedWithLog`. You can see an example in the `View` class used in 2sxc, which provides view-information to the system which is stored as an entity. https://github.com/2sic/2sxc/blob/master/Src/Sxc/ToSic.Sxc/Blocks/Views/View.cs

**Reply by jsboige** [reply]
Thanks for you extended explanation. I'd rather spend more time in VS with type safety so everything I can cram in a dll aside of the 2sxc application folder is welcome. I'm not entirely sure to understand how #2 works then in terms of configuration. How to you tell 2sxc you want to use custom code from a dll?

**Reply by iJungleboy** [reply]
@jsboige I believe it would be better if did a 1-2 hour 1:1 session similar to last time :)

**Reply by jsboige** [reply]
OK sounds good. I'll try to wrap this up and see how far I can get, and I'll get back to you for a live support session.
